docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,421 @@
+| `info` (default) | information about how the control plane operates                                     | user           |
+| `debug`          | extra information about how the control plane operates to help troubleshoot an issue | developer      |
+
+> The levels match the levels of
+> the [controller-runtime framework][cr-log-levels] without any custom levels.
+
+#### Log Messages
+
+We log two kinds of messages (this is the extent of the [logr][logr] API):
+
+- *Error*, which describes a non-critical error in the control plane (for which it will not terminate) that
+  requires the user's attention.
+  > Logging an error is a way of handling it. If you handle the error
+  > differently, log it as info, or don't log it at all.
+  >
+  > See [What not to Log](#what-not-to-log) section about handling critical errors.
+- *Info*, which describes anything that's not an error.
+  - Messages logged at the `info` level target the user persona.
+  - Messages logged at the `debug` level target the developer persona, and we don't expect the
+      user to fully understand them.
+
+> The message kind is not the same as the log level.
+
+#### Context in Messages
+
+In addition to the actual message, the log message should include the context:
+
+- *Timestamp* (handled by the logging library) - to make the user/developer aware *when* something occurs.
+- *Kubernetes resource* (if any) - to correlate to the corresponding Kubernetes resource.
+- *Component name* - to correlate with the component (for example, eventHandler), which logs the message.
+- *Relevant extra context* - additional context depending on the location of the log line in the code.
+
+Additionally, specifically for the developer persona:
+
+- *Stacktrace* - to pinpoint where in the code the message was logged.
+
+The next section shows how to add context.
+
+#### Examples of Log Messages
+
+1. A simple info message with a timestamp with the default `info` level:
+
+   ```go
+   logger.Info("Info message")
+   ```
+
+   ```json
+   {"level":"info","ts":"2023-07-20T15:10:03-04:00","msg":"Info message"}
+   ```
+
+2. An info message with the `debug` level:
+
+   ```go
+    logger.V(1).Info("Info message")
+   ```
+
+   ```json
+   {"level":"debug","ts":"2023-07-20T15:10:03-04:00","msg":"Info message"}
+   ```
+
+3. An error:
+
+   ```go
+    logger.Error(errors.New("test error"), "Error message")
+   ```
+
+   ```json
+    {"level":"error","ts":"2023-07-20T15:10:03-04:00","msg":"Error message","error":"test error","stacktrace":"main.main\n\t/<REDACTED for the document>.go:73\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
+   ```
+
+   > An error message includes the stack trace by default.
+4. A message with a Kubernetes resource:
+
+   ```go
+   // hr implements client.Object (controller-runtime)
+   // hr is *v1beta1.HTTPRoute
+    logger.Info("Processed resource",
```

Comment:
nit: for multiline log messages I think this is preferred according to our style guide:
```suggestion
    logger.Info(
        "Processed resource",
```

Reply:
I found a few violations in the code as well. Will fix those in a separate PR

Reply:
ty!